service/worker/deployment/deployment_reachability.go

```diff
@@ -87,29 +87,55 @@ func getDeploymentReachability(
 	return enumspb.DEPLOYMENT_REACHABILITY_UNREACHABLE, time.Now(), nil
 }
 
+// isDeploymentDrainedOfOpenWorkflows returns true if the deployment contains no open workflows (all pinned workflows are closed)
+func isDeploymentDrainedOfOpenWorkflows(ctx context.Context, nsId, nsName, seriesName, buildId string, visibilityMgr manager.VisibilityManager) (bool, error) {
+	countRequest := makeCountRequest(nsId, nsName, seriesName, buildId, true, true)
+	countResponse, err := visibilityMgr.CountWorkflowExecutions(ctx, &countRequest)
+	if err != nil {
+		return false, err
+	}
+	return countResponse.Count == 0, nil
+}
+
+// isDeploymentDrainedOfAllWorkflows returns true if the deployment contains no open or closed workflows
+func isDeploymentDrainedOfAllWorkflows(ctx context.Context, nsId, nsName, seriesName, buildId string, visibilityMgr manager.VisibilityManager) (bool, error) {
+	countRequest := makeCountRequest(nsId, nsName, seriesName, buildId, false, false)
+	countResponse, err := visibilityMgr.CountWorkflowExecutions(ctx, &countRequest)
+	if err != nil {
+		return false, err
+	}
+	return countResponse.Count == 0, nil
+}
+
 func makeCountRequest(
-	namespaceId, namespaceName, seriesName, buildId string, open bool,
+	namespaceId, namespaceName, seriesName, buildId string, open bool, filterOnStatus bool,
 ) manager.CountWorkflowExecutionsRequest {
 	return manager.CountWorkflowExecutionsRequest{
 		NamespaceID: namespace.ID(namespaceId),
 		Namespace:   namespace.Name(namespaceName),
-		Query:       makeDeploymentQuery(seriesName, buildId, open),
+		Query:       makeDeploymentQuery(seriesName, buildId, open, filterOnStatus),
 	}
 }
 
-func makeDeploymentQuery(seriesName, buildID string, open bool) string {
+func makeDeploymentQuery(seriesName, buildID string, open bool, filterOnStatus bool) string {
```

Comment:
`open bool` and `filterOnStatus bool` together is awkward. better to make it an enum such as:
```
ExecutionStatusFilter {
OpenOnly
ClosedOnly
Any
}
```